versioned/transfer/src/main/java/org/projectnessie/versioned/transfer/CommitLogOptimization.java

```diff
@@ -0,0 +1,261 @@
+              updateCommitsByKeyListSeq.stream().flatMap(Collection::stream))) {
+        entries
+            .map(
+                entry -> {
+                  try {
+                    return databaseAdapter.rebuildKeyList(entry, h -> null);
+                  } catch (ReferenceNotFoundException ignore) {
+                    // ignore
+                    return null;
+                  }
+                })
+            .filter(Objects::nonNull)
+            .forEach(
+                e -> {
+                  try {
+                    databaseAdapter.updateMultipleCommits(singletonList(e));
+                  } catch (ReferenceNotFoundException ex) {
+                    throw new RuntimeException(ex);
+                  }
+                });
+      }
+    }
+  }
+
+  /**
+   * Keeps track of the recent N ({@code parents-per-commit + 1}) commit log entries to populate the
+   * lists of parent commits, both {@link #handleEntry(CommitLogEntry) while scanning} the commit
+   * log and when {@link #drain() reaching the "beginning of time"}.
+   */
+  static final class CommitParentsState {
+
+    private final DatabaseAdapter databaseAdapter;
+    private final int parentsPerCommit;
+    private final BoundedList<CommitLogEntry> lastEntries;
+
+    CommitParentsState(DatabaseAdapter databaseAdapter) {
+      this.databaseAdapter = databaseAdapter;
+      this.parentsPerCommit = databaseAdapter.getConfig().getParentsPerCommit();
+      this.lastEntries = new BoundedList<>(this.parentsPerCommit + 1);
+    }
+
+    boolean canStopIterating(CommitLogEntry entry) {
+      return entry.getParents().size() >= parentsPerCommit;
+    }
+
+    void handleEntry(CommitLogEntry entry) {
+      lastEntries.add(entry);
+      if (lastEntries.size() == parentsPerCommit + 1) {
+        updateLeastRecentCommitLogEntryParents();
+      }
+    }
+
+    void drain() {
+      while (!lastEntries.isEmpty()) {
+        updateLeastRecentCommitLogEntryParents();
+      }
+    }
+
+    private void updateLeastRecentCommitLogEntryParents() {
+      CommitLogEntry commitToUpdate = lastEntries.removeLeastRecent();
+      if (commitToUpdate.getParents().size() >= parentsPerCommit) {
+        return;
+      }
+
+      ImmutableCommitLogEntry.Builder newEntry =
+          ImmutableCommitLogEntry.builder().from(commitToUpdate);
+
+      List<Hash> parents = new ArrayList<>(lastEntries.size());
+      lastEntries.oldestToNewest(e -> parents.add(e.getHash()));
+      if (parents.size() < parentsPerCommit) {
+        parents.add(databaseAdapter.noAncestorHash());
+      }
+
+      List<Hash> currentParents = new ArrayList<>(commitToUpdate.getParents());
+      Hash leastRecentParent = currentParents.get(currentParents.size() - 1);
+      for (int i = 0; i < parents.size(); i++) {
+        if (leastRecentParent.equals(parents.get(i))) {
+          for (i++; i < parents.size(); i++) {
+            currentParents.add(parents.get(i));
```

Comment:
Yes, it's needed. It may happen (as in the `AbstractITExportImportOptimization` in exercise) that parents already exist - maybe just not enough. Added a comment.

Reply:
What I mean is that `parents` represents the list of N immediate parents from the commit log traversal. If `commitToUpdate` has some parents already, I think, those should be contained in `parents` as a sub-list anyway. If they are not in `parents` that would be an error. WDYT?

Then, if `parents` is the most correct version of immediate parents of `commitToUpdate`, we could just use put it into `newEntry.parents` without any extra list manipulation.

Reply:
From a slightly different angle: current code reuses existing parents without validation... Would it make sense to validate them to be a sub-list in `parents`? If so, again we could put `parents` into `newEntry.parents` without building a new list.

Reply:
The parents on the current commit can come from "anywhere".
Technically, you could run commit-log-optimization against an existing repository - after you changed the number of parent-commits for example. So there could be (not enough) parents.

There's much more we could do to improve the cl-optimization - especially batch-writing and less reading.
Wanna say: there's room for optimization for sure, just wonder whether it's worth yet.